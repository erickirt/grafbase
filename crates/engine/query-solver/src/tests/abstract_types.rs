use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
    AGENCY @join__graph(name: "agency", url: "http://localhost:4200/abstract-types/agency")
    BOOKS @join__graph(name: "books", url: "http://localhost:4200/abstract-types/books")
    INVENTORY @join__graph(name: "inventory", url: "http://localhost:4200/abstract-types/inventory")
    MAGAZINES @join__graph(name: "magazines", url: "http://localhost:4200/abstract-types/magazines")
    PRODUCTS @join__graph(name: "products", url: "http://localhost:4200/abstract-types/products")
    REVIEWS @join__graph(name: "reviews", url: "http://localhost:4200/abstract-types/reviews")
    USERS @join__graph(name: "users", url: "http://localhost:4200/abstract-types/users")
}

type Agency
    @join__type(graph: AGENCY, key: "id")
    @join__type(graph: PRODUCTS)
{
    id: ID!
    companyName: String @join__field(graph: AGENCY)
}

type Book implements Product & Similar
    @join__implements(graph: INVENTORY, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Similar")
    @join__implements(graph: REVIEWS, interface: "Product")
    @join__implements(graph: REVIEWS, interface: "Similar")
    @join__type(graph: BOOKS, key: "id")
    @join__type(graph: INVENTORY, key: "id")
    @join__type(graph: PRODUCTS, key: "id")
    @join__type(graph: REVIEWS, key: "id")
{
    id: ID!
    title: String @join__field(graph: BOOKS)
    dimensions: ProductDimension @join__field(graph: PRODUCTS)
    delivery(zip: String): DeliveryEstimates @join__field(graph: INVENTORY, requires: "dimensions { size weight }")
    sku: String @join__field(graph: PRODUCTS)
    createdBy: User @join__field(graph: PRODUCTS)
    similar: [Book] @join__field(graph: PRODUCTS)
    hidden: Boolean @join__field(graph: PRODUCTS)
    publisherType: PublisherType @join__field(graph: PRODUCTS)
    reviewsCount: Int! @join__field(graph: REVIEWS)
    reviewsScore: Float! @join__field(graph: REVIEWS)
    reviews: [Review!]! @join__field(graph: REVIEWS)
    reviewsOfSimilar: [Review!]! @join__field(graph: REVIEWS, requires: "similar { id }")
}

type DeliveryEstimates
    @join__type(graph: INVENTORY)
{
    estimatedDelivery: String
    fastestDelivery: String
}

type Group
    @join__type(graph: AGENCY, key: "id")
{
    id: ID!
    name: String
}

type Magazine implements Product & Similar
    @join__implements(graph: INVENTORY, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Similar")
    @join__implements(graph: REVIEWS, interface: "Product")
    @join__implements(graph: REVIEWS, interface: "Similar")
    @join__type(graph: INVENTORY, key: "id")
    @join__type(graph: MAGAZINES, key: "id")
    @join__type(graph: PRODUCTS, key: "id")
    @join__type(graph: REVIEWS, key: "id")
{
    id: ID!
    dimensions: ProductDimension @join__field(graph: PRODUCTS)
    delivery(zip: String): DeliveryEstimates @join__field(graph: INVENTORY, requires: "dimensions { size weight }")
    title: String @join__field(graph: MAGAZINES)
    sku: String @join__field(graph: PRODUCTS)
    createdBy: User @join__field(graph: PRODUCTS)
    similar: [Magazine] @join__field(graph: PRODUCTS)
    hidden: Boolean @join__field(graph: PRODUCTS)
    publisherType: PublisherType @join__field(graph: PRODUCTS)
    reviewsCount: Int! @join__field(graph: REVIEWS)
    reviewsScore: Float! @join__field(graph: REVIEWS)
    reviews: [Review!]! @join__field(graph: REVIEWS)
    reviewsOfSimilar: [Review!]! @join__field(graph: REVIEWS, requires: "similar { id }")
}

type ProductDimension
    @join__type(graph: INVENTORY)
    @join__type(graph: PRODUCTS)
{
    size: String
    weight: Float
}

type Query
    @join__type(graph: AGENCY)
    @join__type(graph: BOOKS)
    @join__type(graph: INVENTORY)
    @join__type(graph: MAGAZINES)
    @join__type(graph: PRODUCTS)
    @join__type(graph: REVIEWS)
    @join__type(graph: USERS)
{
    books: [Book] @join__field(graph: BOOKS)
    magazines: [Magazine] @join__field(graph: MAGAZINES)
    products: [Product] @join__field(graph: PRODUCTS)
    similar(id: ID!): [Product] @join__field(graph: PRODUCTS)
    review(id: Int!): Review @join__field(graph: REVIEWS)
}

type Review
    @join__type(graph: REVIEWS)
{
    id: Int!
    body: String!
    product: Product
}

type Self
    @join__type(graph: PRODUCTS)
{
    email: String
}

type User
    @join__type(graph: PRODUCTS, key: "email")
    @join__type(graph: USERS, key: "email")
{
    email: ID!
    totalProductsCreated: Int
    name: String @join__field(graph: USERS)
}

interface Product
    @join__type(graph: INVENTORY)
    @join__type(graph: PRODUCTS)
    @join__type(graph: REVIEWS)
{
    id: ID!
    dimensions: ProductDimension @join__field(graph: INVENTORY) @join__field(graph: PRODUCTS)
    delivery(zip: String): DeliveryEstimates @join__field(graph: INVENTORY)
    sku: String @join__field(graph: PRODUCTS)
    createdBy: User @join__field(graph: PRODUCTS)
    hidden: Boolean @inaccessible @join__field(graph: PRODUCTS)
    reviewsCount: Int! @join__field(graph: REVIEWS)
    reviewsScore: Float! @join__field(graph: REVIEWS)
    reviews: [Review!]! @join__field(graph: REVIEWS)
}

interface Similar
    @join__type(graph: PRODUCTS)
    @join__type(graph: REVIEWS)
{
    similar: [Product]
}

union PublisherType
    @join__type(graph: AGENCY)
    @join__type(graph: PRODUCTS)
    @join__unionMember(graph: AGENCY, member: "Agency")
    @join__unionMember(graph: PRODUCTS, member: "Agency")
    @join__unionMember(graph: AGENCY, member: "Group")
    @join__unionMember(graph: PRODUCTS, member: "Self")
 = Agency | Group | Self
"#;

#[test]
fn interface_field_provided_by_implementors() {
    // delivery is not accessible on the Similar interface, but it is on all of its implementors.
    assert_solving_snapshots!(
        "interface_field_provided_by_implementors",
        SCHEMA,
        r#"
        query {
          similar(id: "p1") {
            id
            sku
            delivery(zip: "1234") {
              fastestDelivery
              estimatedDelivery
            }
          }
        }
        "#
    );
}

#[test]
fn nested_interface_field_provided_by_implementors() {
    assert_solving_snapshots!(
        "nested_interface_field_provided_by_implementors",
        SCHEMA,
        r#"
        {
          products {
            id
            reviews {
              product {
                sku
                ... on Magazine {
                  title
                }
                ... on Book {
                  reviewsCount
                }
              }
            }
          }
        }
        "#
    );
}

#[test]
fn unreachable_object() {
    // Group is never reachable.
    assert_solving_snapshots!(
        "unreachable_object",
        SCHEMA,
        r#"
        query {
          products {
            id
            ... on Magazine {
              publisherType {
                ...Publisher
              }
            }
            ... on Book {
              publisherType {
                ...Publisher
              }
            }
          }
        }

        fragment Publisher on PublisherType {
          ... on Agency {
            id
            companyName
          }
          ... on Self {
            email
          }
          ... on Group {
            name
          }
        }
        "#
    );
}
