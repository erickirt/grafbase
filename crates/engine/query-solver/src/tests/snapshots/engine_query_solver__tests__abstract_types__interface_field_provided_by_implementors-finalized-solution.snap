---
source: crates/engine/query-solver/src/tests/abstract_types.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Root#products\", color=royalblue,shape=parallelogram ]\n    2 [ label = \"Query.similar\" ]\n    3 [ label = \"Product.id\" ]\n    4 [ label = \"Product.sku\" ]\n    5 [ label = \"FedEntity#inventory\", color=royalblue,shape=parallelogram ]\n    6 [ label = \"Book.delivery\" ]\n    7 [ label = \"DeliveryEstimates.estimatedDelivery\" ]\n    8 [ label = \"DeliveryEstimates.fastestDelivery\" ]\n    9 [ label = \"*Book.dimensions\" ]\n    10 [ label = \"*ProductDimension.weight\" ]\n    11 [ label = \"*ProductDimension.size\" ]\n    12 [ label = \"*Book.id\" ]\n    13 [ label = \"FedEntity#inventory\", color=royalblue,shape=parallelogram ]\n    14 [ label = \"Magazine.delivery\" ]\n    15 [ label = \"DeliveryEstimates.estimatedDelivery\" ]\n    16 [ label = \"DeliveryEstimates.fastestDelivery\" ]\n    17 [ label = \"*Magazine.dimensions\" ]\n    18 [ label = \"*ProductDimension.weight\" ]\n    19 [ label = \"*ProductDimension.size\" ]\n    20 [ label = \"*Magazine.id\" ]\n    0 -> 1 [ color=royalblue,fontcolor=royalblue ]\n    14 -> 17 [ color=orangered,arrowhead=inv ]\n    2 -> 3 [  ]\n    2 -> 4 [  ]\n    2 -> 5 [ color=royalblue,fontcolor=royalblue ]\n    14 -> 19 [ color=orangered,arrowhead=inv ]\n    6 -> 7 [  ]\n    6 -> 8 [  ]\n    2 -> 9 [  ]\n    9 -> 10 [  ]\n    9 -> 11 [  ]\n    2 -> 12 [  ]\n    2 -> 13 [ color=royalblue,fontcolor=royalblue ]\n    1 -> 2 [  ]\n    14 -> 15 [  ]\n    14 -> 16 [  ]\n    2 -> 17 [  ]\n    17 -> 18 [  ]\n    17 -> 19 [  ]\n    2 -> 20 [  ]\n    5 -> 12 [ color=orangered,arrowhead=inv ]\n    6 -> 10 [ color=orangered,arrowhead=inv ]\n    6 -> 11 [ color=orangered,arrowhead=inv ]\n    6 -> 9 [ color=orangered,arrowhead=inv ]\n    13 -> 20 [ color=orangered,arrowhead=inv ]\n    14 -> 18 [ color=orangered,arrowhead=inv ]\n    13 -> 14 [  ]\n    5 -> 6 [  ]\n}\n"
---
digraph {
    0 [ label = "root" ]
    1 [ label = "Root#products" ]
    2 [ label = "Query.similar" ]
    3 [ label = "Product.id" ]
    4 [ label = "Product.sku" ]
    5 [ label = "FedEntity#inventory" ]
    6 [ label = "Book.delivery" ]
    7 [ label = "DeliveryEstimates.estimatedDelivery" ]
    8 [ label = "DeliveryEstimates.fastestDelivery" ]
    9 [ label = "*Book.dimensions" ]
    10 [ label = "*ProductDimension.weight" ]
    11 [ label = "*ProductDimension.size" ]
    12 [ label = "*Book.id" ]
    13 [ label = "FedEntity#inventory" ]
    14 [ label = "Magazine.delivery" ]
    15 [ label = "DeliveryEstimates.estimatedDelivery" ]
    16 [ label = "DeliveryEstimates.fastestDelivery" ]
    17 [ label = "*Magazine.dimensions" ]
    18 [ label = "*ProductDimension.weight" ]
    19 [ label = "*ProductDimension.size" ]
    20 [ label = "*Magazine.id" ]
    0 -> 1 [ label = "QueryPartition" ]
    14 -> 17 [ label = "RequiredBySubgraph" ]
    2 -> 3 [ label = "Field" ]
    2 -> 4 [ label = "Field" ]
    2 -> 5 [ label = "QueryPartition" ]
    14 -> 19 [ label = "RequiredBySubgraph" ]
    6 -> 7 [ label = "Field" ]
    6 -> 8 [ label = "Field" ]
    2 -> 9 [ label = "Field" ]
    9 -> 10 [ label = "Field" ]
    9 -> 11 [ label = "Field" ]
    2 -> 12 [ label = "Field" ]
    2 -> 13 [ label = "QueryPartition" ]
    1 -> 2 [ label = "Field" ]
    14 -> 15 [ label = "Field" ]
    14 -> 16 [ label = "Field" ]
    2 -> 17 [ label = "Field" ]
    17 -> 18 [ label = "Field" ]
    17 -> 19 [ label = "Field" ]
    2 -> 20 [ label = "Field" ]
    5 -> 12 [ label = "RequiredBySubgraph" ]
    6 -> 10 [ label = "RequiredBySubgraph" ]
    6 -> 11 [ label = "RequiredBySubgraph" ]
    6 -> 9 [ label = "RequiredBySubgraph" ]
    13 -> 20 [ label = "RequiredBySubgraph" ]
    14 -> 18 [ label = "RequiredBySubgraph" ]
    13 -> 14 [ label = "Field" ]
    5 -> 6 [ label = "Field" ]
}
